### PR TITLE
NEW BEPIS: Positronic sphere

### PR DESCRIPTION
**Pull request by Time-Green**
## About The Pull Request

Adds a new positronic brain variant to bepis tech
To save materials we took the corners off but now they can roll
It can roll around, knock items over and push other posibrains away
You can also kick it across the room

https://github.com/tgstation/tgstation/assets/7501474/e4f913e8-cf2f-455a-b2eb-1654468df708

## Why It's Good For The Game
A chorus of whining posibrains is like heroin to the ears, now they can aggresively shit around the robotics lab while whining for their DURAND or cyborg shell, but the robo can also punt them into the disposal bin

## Changelog
:cl:
add: Adds a positronic sphere to bepis tech and roboticist mail goodies. It can now wreack havoc across the robotics lab while whining for a DURAND body, but you can also punt it!
/:cl:

### CODE REVIEW

**Comment by carlarctg** [comment]
can i play soccer wit hti

File: code/game/objects/items/broom.dm [hunk 1]
```diff
@@ -68,10 +68,13 @@
...
  * * A - The atom which is located at the location to push atoms from
  */
-/obj/item/pushbroom/proc/sweep(mob/user, atom/A)
+/obj/item/pushbroom/proc/sweep(mob/user, atom/atom)
 	SIGNAL_HANDLER
 
-	var/turf/current_item_loc = isturf(A) ? A : A.loc
+	do_sweep(src, user, atom, user.dir)
+
+/proc/do_sweep(broomer, mob/user, atom/atom, dir)
```
Comment (MrMelbert) [inline code/game/objects/items/broom.dm:76]: 1. Autodoc
2. Type for `broomer`
3. A more detailed name than `atom/atom` would be nice
4. `dir` could probably be prefixed to be more readable, `sweep_dir`
5. You pass `dir` but `new_item_loc` still gets it from `user.dir`

File: code/modules/mob/living/brain/posibrain.dm [hunk 2]
```diff
@@ -222,3 +222,59 @@ GLOBAL_VAR(posibrain_notify_cooldown)
...
+
+/obj/item/mmi/posibrain/sphere/Moved(atom/old_loc, movement_dir, forced, list/old_locs, momentum_change)
+	. = ..()
+	if(brainmob && isturf(loc))
+		anchored = TRUE
+		do_sweep(src, brainmob, loc, get_dir(old_loc, loc)) //movement dir doesnt work on objects
+		anchored = FALSE
+
+/obj/item/mmi/posibrain/sphere/attack_hand(mob/user, list/modifiers)
+	if(!LAZYACCESS(modifiers, RIGHT_CLICK))
```
Comment (MrMelbert) [inline code/modules/mob/living/brain/posibrain.dm:276]: Use `attack_hand_secondary`

File: code/modules/research/designs/mechfabricator_designs.dm [hunk 3]
```diff
@@ -2605,3 +2605,20 @@
...
+
+/datum/design/posisphere
+	name = "Positronic Sphere"
+	desc = "The latest in Artificial Pesterance."
+	id = "posisphere"
+	build_type = MECHFAB
+	materials = list(
+		/datum/material/iron = HALF_SHEET_MATERIAL_AMOUNT*1.7,
+		/datum/material/glass = HALF_SHEET_MATERIAL_AMOUNT*1.35,
+		/datum/material/gold =SMALL_MATERIAL_AMOUNT*5
```
Comment (MrMelbert) [inline code/modules/research/designs/mechfabricator_designs.dm:2617]:
```suggestion
		/datum/material/iron = HALF_SHEET_MATERIAL_AMOUNT * 1.7,
		/datum/material/glass = HALF_SHEET_MATERIAL_AMOUNT * 1.35,
		/datum/material/gold =SMALL_MATERIAL_AMOUNT * 5,
```

**Review by MrMelbert — CHANGES_REQUESTED** [review]

File: code/modules/research/techweb/all_nodes.dm [hunk 4]
```diff
@@ -2467,3 +2467,30 @@
...
+		"mod_shooting",
+	)
+	research_costs = list(TECHWEB_POINT_TYPE_GENERIC = 2500)
+	hidden = TRUE
+	experimental = TRUE
+
+/datum/techweb_node/posisphere
+	id = "positronic_sphere"
+	display_name = "Experimental Spherical Positronic Brain"
+	description = "Compplaints of bitching posibrains have brought forward a new generation of pestering."
```
Comment (carlarctg) [inline code/modules/research/techweb/all_nodes.dm:2489]:
```suggestion
	description = "Complaints of bitching posibrains have brought forward a new generation of pestering."
```

Comment (carlarctg) [reply code/modules/research/techweb/all_nodes.dm:2489]: alternatively if you take my suggestion on costs

```suggestion
	description = "Recent developments on cost-cutting measures have allowed us to cut positronic brain cubes into cheaper spheres. Unfortunately, it also allows them to move around the lab via rolling maneuvers."
```

Comment (Time-Green) [reply code/modules/research/techweb/all_nodes.dm:2489]: Alright I gotta admit you got some good writing. I'll use it as desc aswell

**Comment by Time-Green** [comment]
> can i play soccer wit hti

yes. you can also play king of the hill on the kitchen table with other posispheres

**Comment by carlarctg** [comment]
i kind of want to suggest them scattering and throwing items they roll over in all directions (instead of pushing and pulling) but that might be too much lol

**Comment by DaCoolBoss** [comment]
Basketball compatability?

File: code/game/objects/items/broom.dm [hunk 5]
```diff
@@ -68,13 +68,17 @@
...
  */
-/obj/item/pushbroom/proc/sweep(mob/user, atom/A)
+/obj/item/pushbroom/proc/sweep(mob/user, atom/atom)
 	SIGNAL_HANDLER
 
-	var/turf/current_item_loc = isturf(A) ? A : A.loc
+	do_sweep(src, user, atom, user.dir)
+
+/// Sweep objects in the direction we're facing towards our direction
+/proc/do_sweep(obj/broomer, mob/user, atom/target, sweep_dir)
```
Comment (MrMelbert) [inline code/game/objects/items/broom.dm:77]: Should include what the args are in the autodoc like typically done

image

File: code/modules/mob/living/brain/posibrain.dm [hunk 6]
```diff
@@ -222,3 +222,62 @@ GLOBAL_VAR(posibrain_notify_cooldown)
...
+	. = ..()
+	if(brainmob && isturf(loc))
+		anchored = TRUE //anchor so we dont broom ourselves.
+		do_sweep(src, brainmob, loc, get_dir(old_loc, loc)) //movement dir doesnt work on objects
+		anchored = FALSE
+
+/// Punt the shit across the room
+/obj/item/mmi/posibrain/sphere/attack_hand_secondary(mob/user, list/modifiers)
+	throw_at(get_edge_target_turf(src, get_dir(user, src)), 7, 1, user)
+	user.do_attack_animation(src)
```
Comment (MrMelbert) [inline code/modules/mob/living/brain/posibrain.dm:283]:
```suggestion
/obj/item/mmi/posibrain/sphere/attack_hand_secondary(mob/user, list/modifiers)
	. = ..()
	if(. == SECONDARY_ATTACK_CANCEL_ATTACK_CHAIN)
		return .
	throw_at(get_edge_target_turf(src, get_dir(user, src)), 7, 1, user)
	user.do_attack_animation(src)
	return SECONDARY_ATTACK_CANCEL_ATTACK_CHAIN
```

**Comment by projectkepler-ru** [comment]
Holy fuck this is some portal 2 type shit, infact
the only thing we're missing is intelligent turret and we're gucci

**Comment by moocowswag** [comment]
GAMESPHERE
image

**Comment by ATH1909** [comment]
is it intentional that items are always knocked south?

**Comment by Time-Green** [comment]
> is it intentional that items are always knocked south?

Nah is fixed

**Comment by Admiral-Gigaton** [comment]
Can you give it inhibited speech, seeing as now it loses some brainpower due to the lack of edges, and becoming cueball smooth

**Comment by DustanHache** [comment]
> Can you give it inhibited speech, seeing as now it loses some brainpower due to the lack of edges, and becoming cueball smooth

you're implying posibrains weren't already smooth. (Also, it could still be rough, it's just that it's round now.(

**Comment by Admiral-Gigaton** [comment]
> > Can you give it inhibited speech, seeing as now it loses some brainpower due to the lack of edges, and becoming cueball smooth
> 
> you're implying posibrains weren't already smooth. (Also, it could still be rough, it's just that it's round now.)

Sure but you just shaved off like... a lot of the cube, which is used to think. It's whatever, just a neat little flavor bit. Would just apply to when it's a cube

File: code/modules/mob/living/brain/posibrain.dm [hunk 7]
```diff
@@ -222,3 +222,64 @@ GLOBAL_VAR(posibrain_notify_cooldown)
...
+
+	if(can_move >= world.time)
+		return
+	can_move = world.time + move_delay
+
+	// ESCAPE PRISON
+	if(ismovable(loc) && prob(25))
+		var/obj/item/item = pick(loc.contents)
+		if(istype(loc, /obj/item/storage))
+			item.forceMove(get_turf(src)) //throw stuff out of the inventory till we free ourselves!
```
Comment (Jacquerel) [inline code/modules/mob/living/brain/posibrain.dm:259]: Can/Should we use drop_location here?

File: code/modules/mob/living/brain/posibrain.dm [hunk 8]
```diff
@@ -222,3 +222,64 @@ GLOBAL_VAR(posibrain_notify_cooldown)
...
+			playsound(src, SFX_RUSTLE, 30, TRUE)
```
Comment (Jacquerel) [inline code/modules/mob/living/brain/posibrain.dm:260]: Maybe a return after the movable check? We don't need to fall back to then checking if we are on a turf

File: code/modules/mob/living/brain/posibrain.dm [hunk 9]
```diff
@@ -222,3 +222,64 @@ GLOBAL_VAR(posibrain_notify_cooldown)
...
+
+	// MOVE US
+	if(isturf(loc))
+		can_move = world.time + move_delay
+		try_step_multiz(direction)
+		SpinAnimation(move_delay, 1, direction == NORTH || direction == EAST)
+
+	return
```
Comment (Jacquerel) [inline code/modules/mob/living/brain/posibrain.dm:268]: Very dilligent, but unecessary

**Review by Jacquerel — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/brain/posibrain.dm [hunk 10]
```diff
@@ -256,17 +256,16 @@ GLOBAL_VAR(posibrain_notify_cooldown)
 	if(ismovable(loc) && prob(25))
 		var/obj/item/item = pick(loc.contents)
 		if(istype(loc, /obj/item/storage))
-			item.forceMove(get_turf(src)) //throw stuff out of the inventory till we free ourselves!
+			item.forceMove(loc.drop_location()) //throw stuff out of the inventory till we free ourselves!
 			playsound(src, SFX_RUSTLE, 30, TRUE)
+			return
```
Comment (Jacquerel) [inline code/modules/mob/living/brain/posibrain.dm:261]:
```suggestion
		return
```
We can return even if you're not in a storage, we've figured out that we're also not a turf.
Still means that 75% of the time that we're in a movable then we will check if we're in a turf anyway but... eh who cares

**Review by Jacquerel — CHANGES_REQUESTED** [review]

**Comment by Stoopler** [comment]
i love this, now we can just have an army of spheres roam the station

**Pull request merged**